### PR TITLE
feat: Add crypto.com provider

### PR DESCRIPTION
**Pull request by rbajollari**
## Description

closes: #1305 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: price-feeder/oracle/provider/crypto.go [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+		Method string            `json:"method"` // public/get-instruments
+		Result CryptoInstruments `json:"result"`
+	}
+	CryptoInstruments struct {
+		Data []CryptoTickerData `json:"data"`
+	}
+	CryptoTickerData struct {
+		InstrumentName string  `json:"i"` // Instrument Name, e.g. BTC_USDT, ETH_CRO, etc.
+		BidPrice       float64 `json:"b"` // The current best bid price, null if there aren't any bids
+		AskPrice       float64 `json:"k"` // The current best ask price, null if there aren't any asks
```
Comment (RafilxTenfen) [inline price-feeder/oracle/provider/crypto.go:114]: Maybe we could remove the fields we don't use ^^

File: price-feeder/oracle/provider/crypto.go [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+		tickerResp    CryptoTickerResponse
+		tickerErr     error
+		candleResp    CryptoCandleResponse
+		candleErr     error
+	)
+
+	// sometimes the message received is not a ticker or a candle response.
+	_ = json.Unmarshal(bz, &heartbeatResp)
+	if heartbeatResp.Method == cryptoHeartbeatMethod {
+		p.pong(bz, reconnectTicker)
```
Comment (RafilxTenfen) [inline price-feeder/oracle/provider/crypto.go:311]: You could pass `heartbeatResp` to the pong func to not unmarshal it again

File: price-feeder/oracle/provider/crypto.go [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+	defer p.mtx.Unlock()
+
+	price, err := coin.NewDecFromFloat(tickerPair.LatestTrade)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker price")
+	}
+	volume, err := coin.NewDecFromFloat(tickerPair.Volume)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker volume")
+	}
```
Comment (RafilxTenfen) [inline price-feeder/oracle/provider/crypto.go:388]:
```suggestion
	price, err := coin.NewDecFromFloat(tickerPair.LatestTrade)
	if err != nil {
		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker price")
		return
	}
	volume, err := coin.NewDecFromFloat(tickerPair.Volume)
	if err != nil {
		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker volume")
		return
	}
```

We should not set if there is a problem with parsing values

File: price-feeder/oracle/provider/crypto.go [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+	defer p.mtx.Unlock()
+
+	close, err := coin.NewDecFromFloat(candlePair.Close)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse candle close")
+	}
+	volume, err := coin.NewDecFromFloat(candlePair.Volume)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse candle volume")
+	}
```
Comment (RafilxTenfen) [inline price-feeder/oracle/provider/crypto.go:407]:
```suggestion
	close, err := coin.NewDecFromFloat(candlePair.Close)
	if err != nil {
		p.logger.Warn().Err(err).Msg("crypto: failed to parse candle close")
		return
	}
	volume, err := coin.NewDecFromFloat(candlePair.Volume)
	if err != nil {
		p.logger.Warn().Err(err).Msg("crypto: failed to parse candle volume")
		return
	}
```

File: price-feeder/oracle/provider/crypto.go [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+	defer p.mtx.Unlock()
+
+	for _, cp := range cps {
+		p.subscribedPairs[cp.String()] = cp
+	}
+}
+
+// GetAvailablePairs returns all pairs to which the provider can subscribe.
+// ex.: map["ATOMUSDT" => {}, "UMEEUSDC" => {}].
+func (p *CryptoProvider) GetAvailablePairs() (map[string]struct{}, error) {
```
Comment (RafilxTenfen) [inline price-feeder/oracle/provider/crypto.go:500]: this endpoint should return an map with the key as `CurrencyPair.String()` which it is "ATOMUSDT", at the moment it is returning "ATOM_USDT"

So, we need to parse it from `CryptoPairsSummary` to `CurrencyPair` and then `availablePairs[currencyPair.Sting()] = struct{}{}`

**Review by RafilxTenfen — CHANGES_REQUESTED** [review]

File: price-feeder/oracle/provider/crypto.go [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+	for i, cp := range cps {
+		pairs[i] = currencyPairToCryptoPair(cp)
+	}
+
+	channels := []string{}
+	for _, pair := range pairs {
+		channels = append(channels, "ticker."+pair)
+	}
+
+	subsMsgID, err := rand.Int(rand.Reader, big.NewInt(1000))
```
[the remark this answers is not in the document]
Comment (rbajollari) [reply price-feeder/oracle/provider/crypto.go:222]: Did it at first thinking the sub id needs to be unique per request but its working now just using 1 as the ID everytime so I've changed it to be 1 all the time.

File: price-feeder/oracle/provider/crypto.go [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+	volume, err := coin.NewDecFromFloat(candlePair.Volume)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse candle volume")
+		return
+	}
+	candle := types.CandlePrice{
+		Price:  close,
+		Volume: volume,
+		// convert seconds -> milli
+		TimeStamp: SecondsToMilli(candlePair.Timestamp),
```
[the remark this answers is not in the document]
Comment (rbajollari) [reply price-feeder/oracle/provider/crypto.go:396]: https://exchange-docs.crypto.com/spot/index.html#candlestick-interval-instrument_name Yeah if you compare the "t" field in the response and the "nonce" in the request which is in unix milliseconds , the example response has "t" in unix seconds.

**Review by RafilxTenfen — APPROVED** [review]
So, the code looks great :100: 

but I was testing with "ATOM_USDT" and received some errors that I think they are okay 
image

But just to note, after some time the crypto provider still was giving 
```shell
2:56PM WRN provider deviating from other candles base=ATOM module=oracle price=12.879940814866863791 provider=crypto
```
cc @adamewozniak

**Pull request merged**